README
= Language.C =

Language.C is a parser and pretty-printer framework for C11 and the extensions of gcc.

See http://visq.github.io/language-c/

== Build and Install ==

cabal install

-- or --

runhaskell Setup.hs configure FLAGS
runhaskell Setup.hs build
runhaskell Setup.hs install

Provide the set of flags passing
 --flags="<flags-seperated-by-space>"
to configure.

== Compatibility ==

Tested with GHC 7.8, 7.10, 8.0 and 8.2.
It is recommended to use the most recent platform release: http://hackage.haskell.org/platform/.

== C Language Compatibility ==

Currently unsupported C11 constructs:
 - static assertion 6.7.10 (_Static_assert)
 - generic selection 6.5.1.1 (_Generic)
 - _Atomic, _Alignas, _Thread_local
 - Universal character names

Currently unsupported GNU C extensions:
 - __auto_type
 - __builtin_offsetof
 - Unicode strings u8"", u""

== Sources ==

see src/README

== Examples ==

A couple of small examples are available in /examples

== Testing ==

A couple of regression tests can be run via
> cd test/harness; make

For more tests, see test/README.


ChangeLog
# Changelog

## WIP

## 0.9.0.2

	- Support parsing an empty list of clobbers after a colon in `asm` statements

## 0.9.0.1

	- Scrap dependency on syb

Thanks to @vmchale

## 0.9
	- Make `Trav` a monad transformer
	- Add `runTravTWithTravState`
	- Tests now use Cabal-3
	- Implement alignment calculation and correct size calculation
	- Add `Show` instance for `Type`
	- Add machine descriptions for x86_64 and armv7l
	- Small correction in output of prettyprinter
	- Allow zeros as line numbers in preprocessed directives

Thanks to @aweinstock314, @ersran9, @expipiplus1, @jkoppel, @trofi and @visq

Changes since 0.7.0
==========================================================================

Wed May 23 07:08:02 2018 +0200
    Support IEC 60559 types and literals. Closes #52.

Wed May 16 07:08:15 2018 +0200
    introduce flag for optional IEC 60559 type support (see #43)

Wed May 16 06:36:04 2018 +0200
    WIP: towards a complete Export (#51)

Thu Feb 22 13:22:50 2018 -0500
    [syntax] GHC.Generics.Generic and Control.DeepSeq.NFData instances (#49)
    (.cabal) drop splitBase flag

Tue Feb 20 08:34:39 2018 +0300
    Selectively hide Prelude.<>:  GHC 8.4 compat (#45)

Tue Feb 20 00:09:16 2018 -0500
    (Analysis) Minimal support for __float128 (#47)

Sat Nov 25 08:59:59 2017 +0100
    Consider builtin_convertvector from #34 for pretty printing

Sat Nov 25 00:46:33 2017 -0600
    add support for _Float128 (#41)

Changes since 0.6.0
==========================================================================
Mon Jul 31 2017
    Improve TypeCheck output (#31)

Sun Jul 30 2017
    * Add `__float128` (fixes #33)

Mon Maz 22 2017
    * Support clang’s __builtin_convertvector (Fixes #34)

Wed Mar 8 2017
    * Move file and parent information to new datatype FilePosition for performance reasons

Changes since 0.5.0
==========================================================================
Thu Feb 23 2017
  * Make gccParseCPPArgs grab the arg to -MF -MT and -MQ
Wed Feb 15 2017
  * In enums, allow multiple attribute specifiers per enumerator
Tue Feb 21 2017
  * Allow typedef redefinition if it denotes the same type
Tue Feb 21 2017
  * Change TypeDefRef to store Type, not Maybe Type
Tue Feb 14 2017
  * Parse (and ignore) Clang __attribute__((availability(id=major.minor.rev)))
Sun Sep 11 2016
  * Add __builtin_bswap32/64.
Wed Jun 22 2016
  * Add '_Alignof' to Lexer.x (fixes #7)
Mon Jun 27 2016
  * Updates for C11 (part 1)
  * _Nullable and _Nonnull support as well as Warnings/lint fixes provided by Anthony Cowley (https://github.com/acowley)
Wed Mar 16 2016
  * Consider storage specifier "ThreadSpec" for global and local declarations
Tue Mar 15 2016
  * Support C11 _NoReturn, genearlize is_inline to FunSpecs (Syntax) / FunAttrs (SemRep)

Changes since 0.4.3
==========================================================================
Wed Mar 2 2016
  * Add direct base type BaseInt128 (complements previous __int128 patch)
    ryan.gl.scott@gmail.com
Sun Feb 28 2016
  * Parse gcc-specific __int128 type
Thu Dec 4 2014
  * Scott Kovach <dskovach@gmail.com>: added derived Eq,Ord instances to NodeInfo

Changes since 0.4.2
==========================================================================
Sat Jan 11 2014
  * Allow unicode characters in string/char literals and filenames
Mon Oct 27 2014
  * macos-attributes
Tue Aug 13 2013
  * Do not derive Error instances for newtypes (type parameter has non-parametric role)
Mon Aug 12 2013
  * Fix bug caused by applying posFile to nopos (reported by Mikhail Sosonkin)

Changes since 0.4.1
==========================================================================
Tue Mar 19 2013
  * TypeCheck: Return Left str instead of fail str (do not rely on MonadError instance of Either)
  * Improve printing of SUERefs and Ident
Thu Feb 28 2013
  * Fix parsing and printing of octal character escapes.
Tue Jun 12 2012
  * Export Annotated type class from AST module
Wed Aug 24 2012
  * Patch for alex-3.0

Changes 0.3.1 - 0.4.1
==========================================================================
Tue Aug 16 2011:
  * Port to ghc-7.2
Fr April 15 2011: Alexander Bernauer <bernauer@inf.ethz.ch>
  * Show instances (popular request) for AST types, DumpAst demo

Changes 0.3 - 0.3.1
==========================================================================
Thu Aug 21 benedikt.huber@gmail.com
  * add aliases for exposed parsers, in order to document them

Fri Aug 15 benedikt.huber@gmail.com
  * Remove NameMap from Data.Name. We will do this right when neccessary.
  * Parser public API: expose parsers and the Parser Monad
  * ParserMonad: Return updated name supply when executing parser
  * Parser: Expose expression, statement, declaration and file parsers
  * Data: Add newNameSupply ~ (namesStartingFrom 0)

Thu Aug 14 17:13:29 CEST 2008  iavor.diatchki@gmail.com
  * Add a utility function to create a "blank" set of cpp arguments.
  * Make that analysis traversal monad abstract.
  * Export the type synonym "Register" (and bump version)

Wed Aug 13 12:00:57 CEST 2008  benedikt.huber@gmail.com
  * add Data.Position: internalIdentAt


Old Changes
==========================================================================

Mon Jun  9 23:12:46 CEST 2008  benedikt.huber@gmail.com
  * License switched to 3-clause BSD
  $
-  In accordance with the original authors, Language.C is now licensed as BSD-3.
   See:
   http://haskell.org/pipermail/c2hs/2008-June/000833.html
   http://haskell.org/pipermail/c2hs/2008-June/000834.html
   http://haskell.org/pipermail/c2hs/2008-June/000835.html



src/Language/C/Data/Ident.hs
{-# LANGUAGE DeriveDataTypeable #-}
-----------------------------------------------------------------------------
-- |
-- Module      :  Language.C.Data.Ident
-- Copyright   :  (c) [1995..1999] Manuel M. T. Chakravarty
--                (c) 2008 Benedikt Huber
-- License     :  BSD-style
-- Maintainer  :  benedikt.huber@gmail.com
-- Stability   :  experimental
-- Portability :  ghc
--
-- This module provides the notion of identifiers in C, speed up using hashing.
-- Identifiers are associated optionally associated with a 'NodeInfo', i.e. with
-- a unique 'Name' and a source location ('Position'). The ordering relation on
-- identifiers is based on the hash and does not follow the lexical order.
-----------------------------------------------------------------------------
module Language.C.Data.Ident (
    Ident(..),
    SUERef(..), isAnonymousRef,
    mkIdent, builtinIdent, internalIdent, internalIdentAt, isInternalIdent,
    identToString, sueRefToString, dumpIdent)
where

-- TODO (comment from manuel):
--  * Hashing is not 8bit clean.

import Data.Char
import Language.C.Data.Position
import Language.C.Data.Node
import Language.C.Data.Name (Name)
import Data.Data (Data)
import Data.Typeable (Typeable)
import GHC.Generics (Generic)
import Control.DeepSeq (NFData)

-- | References uniquely determining a struct, union or enum type.
-- Those are either identified by an string identifier, or by a unique
-- name (anonymous types).
data SUERef =  AnonymousRef Name
             | NamedRef Ident
    deriving (Typeable, Data, Ord, Eq, Show, Generic) --, Read

instance NFData SUERef

-- | Return true if the struct\/union\/enum reference is anonymous.
isAnonymousRef :: SUERef -> Bool
isAnonymousRef (AnonymousRef _) = True
isAnonymousRef _ = False

-- | C identifiers
data Ident = Ident String       -- lexeme
                   {-# UNPACK #-}   !Int     -- hash to speed up equality check
                   NodeInfo                   -- attributes of this ident. incl. position
             deriving (Data,Typeable,Show, Generic) -- Read

instance NFData Ident

-- the definition of the equality allows identifiers to be equal that are
-- defined at different source text positions, and aims at speeding up the
-- equality test, by comparing the lexemes only if the two numbers are equal
--
instance Eq Ident where
  (Ident s h _) == (Ident s' h' _) = (h == h') && (s == s')

-- this does *not* follow the alphanumerical ordering of the lexemes
--
instance Ord Ident where
  compare (Ident s h _) (Ident s' h' _) = compare (h, s) (h', s')

-- identifiers are attributed
instance CNode Ident where
  nodeInfo (Ident _ _ at) = at
instance Pos Ident where
  posOf = posOfNode . nodeInfo
-- to speed up the equality test we compute some hash-like value for each
-- identifiers lexeme and store it in the identifiers representation

-- hash function from the dragon book pp437; assumes 7 bit characters and needs
-- the (nearly) full range of values guaranteed for `Int' by the Haskell
-- language definition; can handle 8 bit characters provided we have 29 bit
-- for the `Int's without sign
--
quad                 :: String -> Int
quad (c1:c2:c3:c4:s)  = ((ord c4 * bits21
                          + ord c3 * bits14
                          + ord c2 * bits7
                          + ord c1)
                         `mod` bits28)
                        + (quad s `mod` bits28)
quad (c1:c2:c3:[]  )  = ord c3 * bits14 + ord c2 * bits7 + ord c1
quad (c1:c2:[]     )  = ord c2 * bits7 + ord c1
quad (c1:[]        )  = ord c1
quad ([]           )  = 0

bits7 :: Int
bits7  = 2^(7::Int)
bits14 :: Int
bits14 = 2^(14::Int)
bits21 :: Int
bits21 = 2^(21::Int)
bits28 :: Int
bits28 = 2^(28::Int)

-- | build an identifier from a string.
--
-- * only minimal error checking, e.g., the characters of the identifier are
--   not checked for being alphanumerical only; the correct lexis of the
--   identifier should be ensured by the caller, e.g., the scanner.
--
-- * for reasons of simplicity the complete lexeme is hashed.
mkIdent            :: Position -> String -> Name -> Ident
mkIdent pos s name  = Ident s (quad s) (mkNodeInfo' pos (pos,length s) name)

-- | returns an /internal/ identifier (has internal position and no unique name)
internalIdent   :: String -> Ident
internalIdent s  = Ident s (quad s) (mkNodeInfoOnlyPos internalPos)

-- | return an /internal/ identifier with position info
internalIdentAt :: Position -> String -> Ident
internalIdentAt pos s = Ident s (quad s) (mkNodeInfoPosLen pos (pos, length s))

-- | returns a /builtin/ identifier (has builtin position and no unique name)
builtinIdent   :: String -> Ident
builtinIdent s  = Ident s (quad s) (mkNodeInfoOnlyPos builtinPos)

-- | return @True@ if the given identifier is /internal/
isInternalIdent :: Ident -> Bool
isInternalIdent (Ident _ _ nodeinfo) = isInternalPos (posOfNode nodeinfo)

-- | string of an identifier
identToString               :: Ident -> String
identToString (Ident s _ _)  = s

-- | string of a SUE ref (empty if anonymous)
sueRefToString                 :: SUERef -> String
sueRefToString (AnonymousRef _) = ""
sueRefToString (NamedRef ident) = identToString ident

-- | dump the identifier string and its positions for debugging purposes
dumpIdent     :: Ident -> String
dumpIdent ide  = identToString ide ++ " at " ++ show (nodeInfo ide)
